concurrent_map.go

```diff
@@ -302,6 +348,40 @@ func (m ConcurrentMap) MarshalJSON() ([]byte, error) {
 	return json.Marshal(tmp)
 }
 
+// isEvictionOccurred check if the map size overflow the maximum allowed size.
+func (m ConcurrentMap) isEvictionOccurred(key string) bool {
+	shard := m.GetShard(key)
+	if m.Count() > shard.maxSize {
+		return true
+	}
+	return false
+}
+
+// removeFirstElement removes the first element from the key slice.
+func (m ConcurrentMap) removeFirstElement(key string) {
+	shard := m.GetShard(key)
+	shard.Lock()
+	copy(shard.mapKeys[0:], shard.mapKeys[1:])
+	shard.mapKeys = shard.mapKeys[:len(shard.mapKeys)-1]
+	shard.Unlock()
+}
+
+// removeMapKey removes the map key from the key slice.
+func (m ConcurrentMap) removeMapKey(key string) {
+	shard := m.GetShard(key)
+
+	if len(shard.mapKeys) > 0 {
+		for i := 0; i < len(shard.mapKeys); i++ {
+			// The mutex lock will be called by the method invoker.
+			if string(key) == shard.mapKeys[i] {
+				// delete key from keys slice
+				copy(shard.mapKeys[i:], shard.mapKeys[i+1:])
```

Comment:
same as my prev comment, copy probably not the best option, we could do some benchmark and see how long it takes when we do copy of 1000 elements over.

If we are using a circular buffer, for removing the element inside an array we could just flag it as "empty"  and not bother as next time the index points to  an "empty location" it is  just replaced.

Reply:
Same as above.